### PR TITLE
Raise an exception when `fetch`ing a stub fails

### PR DESCRIPTION
**Pull request by jonprindiville**
tl;dr I want better error messages for invalid stubs

This is coming out of some discussions that Craig N and I were having
the other day, but this is an issue that has come up a number of times
sporadically in the past.

Today, gapipy swallows some HTTP errors silently (see
`gapipy.query.HTTPERRORS_MAPPED_TO_NONE`). This behaviour kind of makes
sense for the case where you are explicitly `get`ing that resource,
because you can check for that None condition:

    my_departure = g.departures.get(SOME_ID)
    if my_departure is None:
        pass # or do some things

On the other hand, this behaviour is kind of confusing when you're
working with a stub that is embedded in another resource. For example if
I have a `tour_dossiers` instance and I want the name of the corresponding
booking company, I can do something like the following:

    bc_name = my_tour_dossier.booking_companies[0].name

If the reference is correct, this all works fine. If the reference is to
a `booking_companies` that does not exist, or has been deleted or to which
I have no access, though, today gapipy will:

- try and fail to fetch the `booking_companies` by its ID
- try to access the `name` attribute anyway
- throw an AttributeError, claiming that "BookingCompany has no field
name available"

This is of course, a bunch of nonsense: booking companies _do_ have
names, we've simply failed to hydrate a broken reference.

Even if I were to do something like explicitly fetch that reference
before traversing, there's no way for me to see the None that gets
returned, my stub is still a stub.

---

This PR checks for the returned None when `fetch`ing a stub's data, and
raises a more descriptive exception.

NB: I think it could make sense to have some custom exception type for
this instead of using ValueError. Today gapipy doesn't seem to have any
custom exceptions types, though. I'm willing to add it, but figured we
can discuss this idea before ironing that kind of stuff out.

### CODE REVIEW

**Comment by jonprindiville** [comment]
cc @Nagyman

**Comment by Nagyman** [comment]
Very helpful. Thanks for kicking this off!

> NB: I think it could make sense to have some custom exception type for
this instead of using ValueError. Today gapipy doesn't seem to have any
custom exceptions types, though. I'm willing to add it, but figured we
can discuss this idea before ironing that kind of stuff out.

💯 When I was poking around after we discussed this, I considered exception options too. I think there's more flexibility from the client perspective if they're as specific as the underlying HTTP exceptions. That's the problem with swallowing the error and returning `None` in the `get` – the client can't distinguish between forbidden, gone, and never existed. Each of which might be handled differently.

Options...

1. Entirely custom `GapiException(Exception)` ... with more specifics `ResourceGoneException`, `ResourceNotFoundException`, `ResourcePermissionException`
2. Inherit from the HTTP errors `GapiException(HTTPError)` (with similar variations like in 1)
3. Just re-raise the underlying HTTP error

Doing any of this means that the `get` would have to raise the exception, not the `fetch` modified in this PR.

Other thought, `HTTPERRORS_MAPPED_TO_NONE` could be a setting any client could modify to their needs, but that feels a bit weird to me - better to be opinionated and have One Way™ in this case.

(Aside, this change would need a minor version bump)

**Comment by jonprindiville** [comment]
> I think there's more flexibility from the client perspective if they're as specific as the underlying HTTP exceptions. That's the problem with swallowing the error and returning `None` in the `get` – the client can't distinguish between forbidden, gone, and never existed.

Yeah, this is a good point 👍 If we're going to expose errors, we may as well be specific.

> Doing any of this means that the `get` would have to raise the exception, not the `fetch` modified in this PR.

Hm, yeah. Thinking about Django ORM, there the `.get` method can throw `DoesNotExist` and sometimes that works really well with what you're trying to and sometimes it does not. Sometimes we want that _gimme a None instead of an exception_ behaviour and we get it with `.filter(...).first()`, or we use a utility like g.com's `get_object_or_none`.

Echoing that pattern, maybe a potential solution here is:

- `get` should not catch any of those exceptions
- there can be an additional `get_resource_or_none` method someplace that calls `get` but eats the exceptions

Potential downside here is that this is a fairly substantial change to how `get` works -- might be a bunch of code out there that `get`s and then checks for `None` that would all have to change.

> Other thought, `HTTPERRORS_MAPPED_TO_NONE` could be a setting any client could modify to their needs, but that feels a bit weird to me - better to be opinionated and have One Way™ in this case.

Hm, yeah. Alternative to doing it as a setting could be to pass it as an optional kwarg to `get` ... it can default to the current behaviour (_map 403, 404, 410 all to None_) and callers could give something empty/Falsey if they want to let exceptions through.

For example, then in `fetch` if we want the exceptions, instead of doing this `ValueError` stuff, we just pass `httperrors_mapped_to_none=None` or `[]` or whatever.

---

RE exception options: I think that `2` and `3` are better ideas than `1`.

Today it's already possible for gapipy to throw non-403/404/410 `HTTPError`s as well as other `RequestException`-derived stuff. `RequestException` is a base class of `HTTPError`.

File: gapipy/resources/base.py [hunk 1]
```diff
@@ -30,14 +30,17 @@ def options(cls, client):
...
-        self.is_stub = False
 
         # Fetch the resource using the client bound on it, which handles cache get/set.
         resource_obj = getattr(self._client, self._resource_name).get(
             self.id,
-            variation_id=getattr(self, 'variation_id', None))
+            variation_id=getattr(self, 'variation_id', None),
+            httperrors_mapped_to_none=None)
+
         if resource_obj:
```
Comment (jonprindiville) [inline gapipy/resources/base.py]: Hm, maybe this `if` should just go away since we do `httperrors_mapped_to_none=None`... I wonder if there are other circumstances under which this `.get(...)` will return None

**Comment by jonprindiville** [comment]
@Nagyman I've update this to reflect our discussion above. Regarding the exception options, I went with `3. Just re-raise the underlying HTTP error`.

I reworked the unit test as well, though, thinking about it now, maybe it's more straightforward to test that functionality in `test_query` than in `test_resources`... 🤔 

Thoughts?

**Comment by Nagyman** [comment]
@jonprindiville awesome, I like this approach. LGTM. 🚀

**Comment by jonprindiville** [comment]
Here's an example of the behaviour of this branch.

If I have a stub for a departure that my application key cannot see (e.g. using g.com API key and a GEEO departure) and then I access an attribute not in my stub-data...

```
In [14]: stub = gapi.build('departures', {'id': SOME_ID}, stub=True)

In [15]: dstub.sku
```

... a `fetch` happens then we get an error:

```
---------------------------------------------------------------------------
HTTPError                                 Traceback (most recent call last)

[... traceback removed ...]

HTTPError: 403 Client Error: {"http_status_code":403,"message":"You do not have permission to perform this action.","errors":[],"time":"2019-12-05T15:55:46Z","error_id":"gapi_643efdd09d7b4e60b9f712118891b4ab"}
```

In this case, 403, because it's a permissions issue. (Compared with current gapipy, which would give you an `AttributeError` saying sku is not an available field)

File: gapipy/resources/base.py [hunk 2]
```diff
@@ -30,14 +30,16 @@ def options(cls, client):
...
         resource_obj = getattr(self._client, self._resource_name).get(
             self.id,
-            variation_id=getattr(self, 'variation_id', None))
-        if resource_obj:
-            self._fill_fields(resource_obj._raw_data)
+            variation_id=getattr(self, 'variation_id', None),
+            httperrors_mapped_to_none=None)
+
+        self._fill_fields(resource_obj._raw_data)
+        self.is_stub = False
```
Comment (marz619) [inline gapipy/resources/base.py]: 👍

**Comment by marz619** [comment]
> I've update this to reflect our discussion above. Regarding the exception options, I went with `3. Just re-raise the underlying HTTP error`.

Cool approach! This doesn't change the raise an exception behaviour from `Resource.__getattr__`. It returns more _expressive_ exceptions that better quantify the underlying issue.

> I reworked the unit test as well, though, thinking about it now, maybe it's more straightforward to test that functionality in `test_query` than in `test_resources`... 🤔

Since this is preserving the existing behaviour of the `Resource.fetch`, and modifying the `Query.get` functionality, I agree it makes sense to put the test in `test_query`.

```python
""" preserved behaviour """

>>> gapi.departures.get('1234567').sku
# ---------------------------------------------------------------------------
# AttributeError                            Traceback (most recent call last)
# ----> gapi.departures.get('1234567').sku
#
# AttributeError: 'NoneType' object has no attribute 'sku'

""" new functionality """

>>> gapi.departures.get('1234567', httperrors_mapped_to_none=None).sku
# ---------------------------------------------------------------------------
# HTTPError                                 Traceback (most recent call last)
#
# [ ... traceback removed ...]
#
# HTTPError: 404 Client Error: {"http_status_code":404,"message":"Not found.","errors":[],"time":"2019-12-12T23:45:35Z","error_id":"gapi_54afe219083143488324cf153f848e45"} for url: https://rest.gadventures.com/departures/1234567
```

This preservation of existing behaviour would possibly lead to rewriting a lot of existing code to include the new `httperrors_mapped_to_none` kwarg everywhere a `Query.get` is being used in order to access the more relevant HTTPError's. But maybe that's a discussion for a future `gapipy v3` ¯\_(ツ)_/¯?

@jonprindiville one final note, if you could rebase this from the latest `master` ... I'm happy to merge and deploy this as `2.25.0`

**Comment by jonprindiville** [comment]
> > I've update this to reflect our discussion above. Regarding the exception options, I went with `3. Just re-raise the underlying HTTP error`.
> 
> Cool approach! This doesn't change the raise an exception behaviour from `Resource.__getattr__`. It returns more _expressive_ exceptions that better quantify the underlying issue.

Yeah, I could understand interpreting it as new behaviour, though, I guess. RequestException-based throws in addition to AttributeError. (I imagine it's still possible to get an AttributeError if you actually supply a non-existent attr.)

> Since this is preserving the existing behaviour of the `Resource.fetch`, and modifying the `Query.get` functionality, I agree it makes sense to put the test in `test_query`.

👍 works for me, will move it (I think moving it might make it easier to test, actually)

> This preservation of existing behaviour would possibly lead to rewriting a lot of existing code to include the new `httperrors_mapped_to_none` kwarg everywhere a `Query.get` is being used in order to access the more relevant HTTPError's.

Yeah... I was a bit on the fence about changing the default behaviour of `get` or not. I figured **not** changing the default behaviour of `get` would be easier to get merged. 🙃 

In terms of breaking existing code... I doubt folks are catching AttributeError from their `get` attempts as a way to detect 403/404/410, but I guess they _could_ be.

>  But maybe that's a discussion for a future `gapipy v3` ¯_(ツ)_/¯?

Yeah... IDK if there's any kind of roadmap. Whenever we _do_ get to bumping to v3, maybe it's worth changing the default behaviour of get to throw exceptions. 🤷‍♂ 
 
> @jonprindiville one final note, if you could rebase this from the latest `master` ... I'm happy to merge and deploy this as `2.25.0`

👍 sure thing

**Comment by jonprindiville** [comment]
@marz619 this has been rebased on latest master, and I moved the testing out of `test_resources` and into `test_query` since most of the actual changes were to the Query class.

LMK if you have any more thoughts here.

You think it's worth mentioning this `httperrors_mapped_to_none` in the README? There is the briefest of descriptions there about how to use the Query objects.

(Please squash when merging... there's definitely some experimenty/back-and-forthy kind of commits here)
